src/util/hash.rs

```diff
@@ -25,48 +25,50 @@ use core::cmp::min;
 use hashes::Hash;
 use consensus::encode::Encodable;
 
-/// Calculates the merkle root of a list of hashes inline
-/// into the allocated slice.
+/// Calculates the merkle root of a list of hashes inline (in place) in `data`.
 ///
 /// In most cases, you'll want to use [bitcoin_merkle_root] instead.
+///
+/// # Panics
+/// Panics if, and only if, `data` is empty. This function should only ever be called with the
+/// transaction data from a block, which should never be empty because blocks always contain at
+/// least one transaction (the coinbase transaction).
 pub fn bitcoin_merkle_root_inline<T>(data: &mut [T]) -> T
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
 {
-    // Base case
-    if data.is_empty() {
-        return Default::default();
-    }
-    if data.len() < 2 {
-        return T::from_inner(data[0].into_inner());
+    if data.len() == 0 {
+        panic!("merkle root of an empty tree is undefined");
     }
-    // Recursion
-    for idx in 0..((data.len() + 1) / 2) {
-        let idx1 = 2 * idx;
-        let idx2 = min(idx1 + 1, data.len() - 1);
-        let mut encoder = T::engine();
-        data[idx1].consensus_encode(&mut encoder).unwrap();
-        data[idx2].consensus_encode(&mut encoder).unwrap();
-        data[idx] = T::from_engine(encoder);
+
+    // If there is only a single hash then that hash is, by definition, the merkle root.
+    if data.len() == 1 {
+        return data[0];
```

Comment:
This same condition is in `markle_root_r`, so probably not needed here.

Reply:
I added it there for two reasons:
1. It gives a line of code that can carry the code comment
2. In `merkle_root_r` this call is the recursion base case, here it is not, so slightly different

I was trying to separate the recursion from the two 'higher' functions.